src/libclipper/src/selection_policies.cpp

```diff
@@ -1,344 +1,484 @@
-          // means this model predicted the right label
-          model_weight.second *= 1.1;
-          weight_sum += model_weight.second;
-        } else {
-          // means this model predicted the wrong label
-          model_weight.second *= 0.9;
-          weight_sum += model_weight.second;
-        }
-        break;
+std::string EpsilonGreedyPolicy::serialize_state(PolicyState state) {
+  return state.serialize();
+}
+
+PolicyState EpsilonGreedyPolicy::deserialize_state(const std::string& bytes) {
+  return PolicyState::deserialize(bytes);
+}
+
+std::string EpsilonGreedyPolicy::state_debug_string(const PolicyState& state) {
+  return state.debug_string();
+};
+
+// ********
+// * UCB1 *
+// ********
+
+PolicyState UCBPolicy::initialize(
+    const std::vector<VersionedModelId>& candidate_models_) {
+  return EpsilonGreedyPolicy::initialize(candidate_models_);
+}
+
+PolicyState UCBPolicy::add_models(
+    PolicyState state, const std::vector<VersionedModelId>& new_models) {
+  return EpsilonGreedyPolicy::add_models(state, new_models);
+}
+
+VersionedModelId UCBPolicy::select(PolicyState& state) {
+  
+  // Helper function for selecting an arm based on lowest upper confidence bound
+  VersionedModelId selected_model;
+  if (state.model_map_.empty()) { //Edge case
+    std::cout << "No models to select from." << std::endl;
+  } else {
+    auto min_upper_bound = DBL_MAX;
+    for (auto id = state.model_map_.begin(); id != state.model_map_.end(); ++id) {
+      auto model_loss = id->second["expected_loss"];
+      auto bound = sqrt(2 * log(state.model_map_.size()) / id->second["times_selected"]);
+      if (model_loss + bound < min_upper_bound) {
+        min_upper_bound = model_loss + bound;
+        selected_model = id->first;
       }
     }
   }
 
-  for (std::pair<VersionedModelId, float>& model_weight : state) {
-    model_weight.second /= weight_sum;
-  }
-  return state;
+  return selected_model;
 }
 
-std::string BanditPolicy::serialize_state(BanditState state) {
-  std::ostringstream ss;
-  // TODO: clean up this function. The nested pairs make it hard to read.
-  for (auto m = state.begin(); m != state.end() - 1; ++m) {
-    ss << m->first.first << ITEM_PART_CONCATENATOR << m->first.second
-       << ITEM_PART_CONCATENATOR << std::to_string(m->second) << ITEM_DELIMITER;
-  }
-  // don't forget to save the last label
-  ss << (state.end() - 1)->first.first << ITEM_PART_CONCATENATOR
-     << (state.end() - 1)->first.second << ITEM_PART_CONCATENATOR
-     << std::to_string((state.end() - 1)->second);
-  std::cout << "BanditPolicy::serialize_state result: " << ss.str()
-            << std::endl;
-  return ss.str();
+std::vector<PredictTask> UCBPolicy::select_predict_tasks(PolicyState& state,
+                                                         Query query,
+                                                         long query_id) {
+  return Exp3Policy::select_predict_tasks(state, query, query_id);
```

Comment:
Why are you using `Exp3Policy` to select? Shouldn't you be using `UCBPolicy::select`?